worlds/zillion/__init__.py

```diff
@@ -0,0 +1,374 @@
+
+    It's based on the anime Zillion (赤い光弾ジリオン, Akai Koudan Zillion).
+    """
+    game = "Zillion"
+    web = ZillionWebWorld()
+
+    option_definitions: Dict[str, AssembleOptions] = zillion_options
+    topology_present: bool = True  # indicate if world type has any meaningful layout/pathing
+
+    # gets automatically populated with all item and item group names
+    all_item_and_group_names: FrozenSet[str] = frozenset()
+
+    # map names to their IDs
+    item_name_to_id: Dict[str, int] = _item_name_to_id
+    location_name_to_id: Dict[str, int] = _loc_name_to_id
+
+    # increment this every time something in your world's names/id mappings changes.
+    # While this is set to 0 in *any* AutoWorld, the entire DataPackage is considered in testing mode and will be
+    # retrieved by clients on every connection.
+    data_version: int = 1
+
+    # NOTE: remote_items and remote_start_inventory are now available in the network protocol for the client to set.
+    # These values will be removed.
+    # if a world is set to remote_items, then it just needs to send location checks to the server and the server
+    # sends back the items
+    # if a world is set to remote_items = False, then the server never sends an item where receiver == finder,
+    # the client finds its own items in its own world.
+    remote_items: bool = False
+
+    logger: logging.Logger
+
+    id_to_zz_item: Optional[Dict[int, ZzItem]] = None
+    zz_system: System
+    _item_counts: "Counter[str]" = Counter()
+    """
+    These are the items counts that will be in the game,
+    which might be different from the item counts the player asked for in options
+    (if the player asked for something invalid).
+    """
+    my_locations: List[ZillionLocation] = []
+    """ This is kind of a cache to avoid iterating through all the multiworld locations in logic. """
+
+    def __init__(self, world: MultiWorld, player: int):
+        super().__init__(world, player)
+        self.logger = logging.getLogger("Zillion")
+        self.zz_system = System()
+        clear_cache()
+
+    def _make_item_maps(self, start_char: Chars) -> None:
+        _id_to_name, _id_to_zz_id, id_to_zz_item = make_id_to_others(start_char)
+        self.id_to_zz_item = id_to_zz_item
+
+    @classmethod
+    def stage_assert_generate(cls, world: MultiWorld) -> None:
+        """Checks that a game is capable of generating, usually checks for some base file like a ROM.
+        Not run for unittests since they don't produce output"""
+        rom_file = get_base_rom_path()
+        if not os.path.exists(rom_file):
+            raise FileNotFoundError(rom_file)
+
+    def generate_early(self) -> None:
+        zz_op, item_counts = validate(self.world, self.player)
+
+        self._item_counts = item_counts
+
+        rom_dir_name = os.path.dirname(get_base_rom_path())
+        self.zz_system.make_patcher(rom_dir_name)
```

Comment:
If possible, everything output-related should be run in generate_output since that runs multi-threaded. If you want to throw early exceptions, that's what assert_generate is for.

Reply:
This "patcher" has some other stuff besides patching. Probably less-than-optimal organization in my package.
This is just initializing the instance to be available for those other things.